### PR TITLE
Fix signature of overrides from QAbstractItemModel

### PR DESCRIPTION
**Pull request by eivindjahren**
To satisfy LSP it should have exactly the same signature as QAbstractItemModel.

- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Added appropriate release note label
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [x] Make sure tests pass locally (after every commit!)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: src/ert/gui/ertwidgets/models/storage_model.py [hunk 1]
```diff
@@ -27,6 +29,8 @@ class _Column(IntEnum):
     2: "Type",
 }
 
+default_index = QModelIndex()
```
Comment (andreas-el) [inline src/ert/gui/ertwidgets/models/storage_model.py:32]: Out of curiosity, is the reason for this the same as with dicts?

File: src/ert/gui/ertwidgets/models/storage_model.py [hunk 2]
```diff
@@ -165,45 +169,59 @@ def _load_storage(self, storage: Storage) -> None:
...
 
-    def parent(self, index: QModelIndex) -> QModelIndex:
-        if not index.isValid():
+    @overload
+    def parent(self, child: QModelIndex) -> QModelIndex: ...
+    @overload
+    def parent(self) -> Optional[QObject]: ...
+    @override
+    def parent(self, child: QModelIndex = default_index) -> Optional[QObject]:
+        if not child.isValid():
```
Comment (andreas-el) [inline src/ert/gui/ertwidgets/models/storage_model.py:191]: Why does this function return Optional[QObject] ?
As far as I can see this returns a QModelIndex on all exits?

File: src/ert/gui/ertwidgets/models/storage_model.py (hunk 1 above)
Comment (eivindjahren) [reply src/ert/gui/ertwidgets/models/storage_model.py:32]: So this explictly sets up the same thing as `parent=QModelIndex()`. There will only be one copy. Maybe we should just call it in the parameter? I remember there was a time where some linting rule said to just not do that.

Comment (eivindjahren) [reply src/ert/gui/ertwidgets/models/storage_model.py:32]: I will try to redo this in the `if parent is None:` way and see if mypy complains.

Comment (eivindjahren) [reply src/ert/gui/ertwidgets/models/storage_model.py:32]: Yea, this was a mistake, I should have used an Optional here as it is variant.

Comment (eivindjahren) [reply src/ert/gui/ertwidgets/models/storage_model.py:32]: Fixed it now.

**Review by andreas-el — APPROVED** [review]
Fine work. 👍

File: src/ert/gui/ertwidgets/models/storage_model.py (hunk 2 above)
Comment (eivindjahren) [reply src/ert/gui/ertwidgets/models/storage_model.py]: That is true, but the type checker does not like that. I believe this has to do with overloads as a similar situation works correctly when there is no overload involved.

Overloads are somewhat of a special case. Here you would have to figure out that

```python
    def parent(self, child: Optional[QModelIndex] = None) -> QModelIndex:
```

satisfies the following interface:

```python
    @overload
    def parent(self, child: QModelIndex) -> QModelIndex: ...
    @overload
    def parent(self) -> Optional[QObject]: ...
```

which is quite complicated when you take into account all possibilities. mypy will not check that you always get a QModelIndex when the parameter `child` is supplied for instance, but it does check that the signature for the implementation accepts all the parameters in the overload defintions.

**Pull request merged**